src/Stratis.Bitcoin/Connection/PeerLifetime.cs

```diff
@@ -0,0 +1,109 @@
+﻿using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Utilities;
+using System;
+using System.Collections.Concurrent;
+using System.Net;
+
+namespace Stratis.Bitcoin.Connection
+{
+    /// <summary>
+    /// A class that will manage the lifetime of connected peers and be in charge of banning peers that are malicious.
+    /// </summary>
+    public interface IPeerLifetime
+    {
+        void BanPeer(IPEndPoint endPoint, string reason = null);
+
+        bool IsBanned(IPEndPoint endPoint);
+    }
+
+    /// <summary>
+    /// A class that will manage the lifetime of connected peers and be in charge of banning peers that are malicious.
+    /// </summary>
+    public interface IBanStore
+    {
+        void BanPeer(IPEndPoint endPoint, DateTime banUntil);
+
+        DateTime? TryGetBanedPeer(IPEndPoint endPoint);
+    }
+
+    /// <summary>
+    /// A temporary memory store for banned peers.
+    /// </summary>
+    public class MemoryBanStore : IBanStore
+    {
+        private readonly ConcurrentDictionary<IPAddress, DateTime> banned = new ConcurrentDictionary<IPAddress, DateTime>();
+
+        public void BanPeer(IPEndPoint endPoint, DateTime banUntil)
+        {
+            this.banned.AddOrUpdate(endPoint.Address, banUntil, (point, time) => banUntil);
+        }
+
+        public DateTime? TryGetBanedPeer(IPEndPoint endPoint)
+        {
+            if (this.banned.TryGetValue(endPoint.Address, out DateTime banUntil))
+                return banUntil;
+
+            return null;
+        }
+    }
+
+    public class PeerLifetime : IPeerLifetime
+    {
+        /// <summary>A connection manager of peers.</summary>
+        private readonly IConnectionManager connectionManager;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>A store of banned peers.</summary>
+        private readonly IBanStore banStore;
+
+        /// <summary>Functionality of date and time.</summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        public PeerLifetime(IConnectionManager connectionManager, ILoggerFactory loggerFactory, IDateTimeProvider dateTimeProvider)//, IBanStore banStore)
+        {
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.connectionManager = connectionManager;
+            this.banStore =  new MemoryBanStore();// banStore;
+            this.dateTimeProvider = dateTimeProvider;
+        }
+
+        /// <inheritdoc />
+        public void BanPeer(IPEndPoint endPoint, string reason = null)
+        {
+            Guard.NotNull(endPoint, nameof(endPoint));
+
+            this.logger.LogTrace("()");
+
+            this.banStore.BanPeer(endPoint, this.dateTimeProvider.GetUtcNow().AddDays(7));
+
+            var peer = this.connectionManager.ConnectedNodes.FindByEndpoint(endPoint);
+            if (peer != null)
```

Comment:
I don't really understand the logic here.
So this peer has been banned, but then we check if it's in the whitelist to disconnect. Should you not always disconnect when you ban a peer?

Reply:
Whitelisted are peers we configured to connect to `addnode` they are normally trusted peers.
But you are right I changed it so we don't ban whitelisted peers either.